### DISCUSSION TITLE
Interrupt running script

### DISCUSSION THREAD

**Discussion by kotakotik22** [mozilla/rhino#1154] [Q&A] [answered]
im trying to make it possible to stop a script running in its own thread, but i cant find any way to do it
im testing with this script:
```js
while(true) {
    try {
        sleep(1000)
        print("Still running")
    } catch(e) {
        print(`Catched ${e}`)
    }
}
```
i tried exceptions, but the catch block catches any exception or error
thread interrupt also uses an exception so that doesnt help in any way

and i tried closing the scope (with LOTS of reflection because the built-in function closes the context in the thread) but it doesnt seem to do anything (no errors, script keeps running) so im guessing it doesnt do anything to the script

**Comment by rPraml** [comment]
You may take a look at the continuation API https://github.com/mozilla/rhino/blob/master/testsrc/org/mozilla/javascript/tests/ContinuationsApiTest.java or the debugger API. The debugger can set breakpoints and pause the script.

To open the debugger you may use this code snippet

```java
import org.mozilla.javascript.tools.debugger.Main;
...
Context cx = Context.enter();
try {
  ScopeProvider provider = () -> ScriptRuntime.getTopCallScope(Context.getCurrentContext());
  Main main = Main.mainEmbedded(cx.getFactory(), provider, "Debugger");
  main.setExitAction(() -> main.detach());
} finally {
  Context.exit();
}
```

**Reply by kotakotik22** [reply]
the debugger opens a window and im trying to stop the script from my own signal
``captureContinuation`` causes an error because its trying to create an object in the context associated with the current thread and im trying to stop the script from another thread

**Comment by p-bakker** [comment, marked as answer]
@kotakotik22 from where are you trying to interrupt the script? From inside the script itself of from the outside, in Java code?

Are you running in interpreted mode or compiled? Also see https://github.com/mozilla/rhino/issues/49 and https://github.com/mozilla/rhino/issues/175

**Reply by kotakotik22** [reply]
im trying to stop the script from java, and i think im running compiled, im just using the evaluateString method like the [RunScript.java](https://github.com/mozilla/rhino/blob/6c3a17faadf6adda6c37a9a48546fd245099b2d7/examples/RunScript.java#L37) example

**Reply by kotakotik22** [reply]
using ``observeInstructionCount`` works
i check every 1 instruction whether the script should be stopped and if it is, i throw an exception
